server/py/services/api/migrations/env.py

```diff
@@ -112,6 +113,32 @@ def compare_type(
     return None
 
 
+@contextlib.contextmanager
+def _get_connection():
+    connection_or_engine = alembic.context.config.attributes.get("connection")
+
+    if connection_or_engine is None:
+        engine = sqlalchemy.engine_from_config(
+            config.get_section(config.config_ini_section),
+            prefix="sqlalchemy.",
+            poolclass=sqlalchemy.pool.NullPool,
+        )
+        with engine.connect() as conn:
+            yield conn
+        return
+
+    if isinstance(connection_or_engine, sqlalchemy.engine.Engine):
+        with connection_or_engine.connect() as conn:
+            yield conn
+        return
+
+    if isinstance(connection_or_engine, sqlalchemy.engine.Connection):
+        yield connection_or_engine
+        return
```

Comment:
can you elaborate on which flows we can engine and cases where we get connection?